### PR TITLE
UCM: Avoid using thread-local-storage as recursion guard.

### PR DESCRIPTION
**Pull request by yosefe**

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1466/ for details.

**Comment by yosefe** [comment]
@alex--m this should be the bug you had today

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/861/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1467/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/862/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
@shamisp  @miked-mellanox pls review

File: src/ucm/mmap/replace.c [hunk 1]
```diff
@@ -16,19 +16,21 @@
...
 #include <ucs/sys/preprocessor.h>
 #include <ucs/type/component.h>
-#include <dlfcn.h>
+#include <pthread.h>
 
 
 #define MAP_FAILED ((void*)-1)
 
-static __thread int ucm_mmap_reentry = 0;
+static pthread_mutex_t ucm_mmap_get_orig_lock = PTHREAD_RECURSIVE_MUTEX_INITIALIZER_NP;
```
Comment (miked-mellanox) [inline src/ucm/mmap/replace.c]: out of interest, why thread local storage caused  a problem?

Comment (yosefe) [inline src/ucm/mmap/replace.c]: because thread local storage implementation called malloc(), which resulted in a recursive invocation of sbrk()

**Comment by shamisp** [comment]
:+1: Looks good to me

**Pull request merged**